crates/bevy_ecs/src/query/iter.rs

```diff
@@ -69,73 +46,197 @@ where
+                            &self.archetypes,
+                            &self.query_state,
+                        ) {
+                            Some(_) => {}
+                            None if i > 0 => continue 'outer,
+                            None => return None,
+                        }
+                    }
+                    break;
                 }
+                None if i > 0 => continue,
+                None => return None,
             }
         }
+
+        // MaybeUninit::uninit_array is unstable
+        let mut values: [MaybeUninit<<Q::Fetch as Fetch<'a>>::Item>; K] =
+            MaybeUninit::uninit().assume_init();
+
+        for (value, cursor) in values.iter_mut().zip(&mut self.cursors) {
+            value.as_mut_ptr().write(cursor.peek_last().unwrap());
+        }
+
+        // MaybeUninit::array_assume_init is unstable
+        let values: [<Q::Fetch as Fetch<'a>>::Item; K] =
+            (&values as *const _ as *const [<Q::Fetch as Fetch<'a>>::Item; K]).read();
+
+        Some(values)
+    }
+
+    /// Get next combination of queried components
+    pub fn fetch_next(&mut self) -> Option<[<Q::Fetch as Fetch<'_>>::Item; K]>
+    where
+        Q::Fetch: Clone,
+        F::Fetch: Clone,
+    {
+        // safety: we are limiting the returned reference to self,
+        // making sure this method cannot be called multiple times without getting rid
+        // of any previously returned unique references first, thus preventing aliasing.
+        unsafe { self.fetch_next_aliased_unchecked() }
+    }
+}
+
+// Iterator type is intentionally implemented only for read-only access.
+// Doing so for mutable references would be unsound, because
+// calling `next` multiple times would allow
+impl<'w, 's, Q: WorldQuery, F: WorldQuery, const K: usize> Iterator
+    for QueryCombinationIter<'w, 's, Q, F, K>
+where
+    Q::Fetch: Clone + ReadOnlyFetch,
+    F::Fetch: Clone + FilterFetch + ReadOnlyFetch,
+{
+    type Item = [<Q::Fetch as Fetch<'w>>::Item; K];
+
+    #[inline]
+    fn next(&mut self) -> Option<Self::Item> {
+        // Safety: it is safe to alias for ReadOnlyFetch
+        unsafe { QueryCombinationIter::fetch_next_aliased_unchecked(self) }
     }
 
     // NOTE: For unfiltered Queries this should actually return a exact size hint,
     // to fulfil the ExactSizeIterator invariant, but this isn't practical without specialization.
     // For more information see Issue #1686.
     fn size_hint(&self) -> (usize, Option<usize>) {
-        let max_size = self
+        if K == 0 {
+            return (0, Some(0));
+        }
+
+        let max_size: usize = self
             .query_state
             .matched_archetypes
             .ones()
             .map(|index| self.world.archetypes[ArchetypeId::new(index)].len())
             .sum();
 
-        (0, Some(max_size))
+        if max_size < K {
+            return (0, Some(0));
+        }
+
+        // n! / k!(n-k)! = (n*n-1*...*n-k+1) / k!
```

Comment:
IIRC, `n! / k!(n-k)!` is a formula for combinations, not permutations.
`n! / (n-k)!` is the formula for permutations.

Reply:
```rust
let permutations = |n, k| (n - k + 1..=n).product();
```

Reply:
This is an iterator of combinations, not permutations. Also it is verified in tests.

Reply:
I was just mostly confused because you gave the variable the name `max_permutations` and yet the algorithm for it calculates combinations.

Reply:
Oh, right. That was my own confusion about that going through. I've renamed the methods, but forgot about the local variables.